### ISSUE TITLE
query the image digest of a tag from the registry

### ISSUE THREAD

**Issue by aespinosa** [GoogleContainerTools/jib#2034]
**Environment**:

- *Jib version:* jib-core 0.11.0
- *Build tool:* maven 3.6
- *OS:* macos

**Description of the issue**:

Looking at https://github.com/GoogleContainerTools/jib/blob/master/jib-core/src/main/java/com/google/cloud/tools/jib/registry/ManifestPuller.java#L89-L94 the ManifestPuller doesn't read the response header of `Docker-Content-Digest` to get the sha256 digest you can use and pass to tools like `docker pull <image-name>@sha256:....` or `kubectl set image deployment/name container=<image-name>@sha256...` for CI/ CD purposes. 

**Expected behavior**:

A new method `getDigest()` from the puller

**Steps to reproduce**:
Run the following test:

```
public class RegistryTest {
    @Test
    public void registryDigest() throws IOException, RegistryException {
        RegistryClient.Factory factory = RegistryClient.factory(EventHandlers.NONE, "registry-1.docker.io", "library/centos");
        Authorization auth = factory.newRegistryClient().getRegistryAuthenticator().get().authenticatePull(null);
        factory.setAuthorization(auth);
        RegistryClient client = factory.newRegistryClient();

        V22ManifestListTemplate manifest = client.pullManifest("7", V22ManifestListTemplate.class);
        // If this gets stale, update by getting the output of
        // skopeo  inspect docker://centos:7 | jq '.Digest' -r
        Assertions.assertEquals("sha256:307835c385f656ec2e2fec602cf093224173c51119bbebd602c53c3653a3d6eb", manifest.getDigest());

        // These refer to other digests found in the manifest.json
//        Assertions.assertEquals("foo", manifest.getDigestsForPlatform("amd64", "linux"));
//        V22ManifestTemplate manifest = client.pullManifest("7", V22ManifestTemplate.class);
//        Assertions.assertEquals("foo", manifest.getContainerConfiguration().getDigest());
    }
}
```

**Log output**: 

none, the code doesn't compile

**Comment by chanseokoh** [comment]
Hi @aespinosa,

Allowing to get the manifest digest makes sense. Actually, we were recently discussing some ways to get the hash, and we will probably re-compute the digest on our end instead of using `Docker-Content-Digest` (https://github.com/GoogleContainerTools/jib/issues/1884#issuecomment-537160441).

As for the API, I don't think we will add a method like `getDigest()`, as the contract should be that the class is immutable. Rather, I think it should return both digest and template.

```java
/**
 * Provides implementations for a registry endpoint. Implementations should be immutable.
   ...
 */
interface RegistryEndpointProvider<T> {
```

**Comment by aespinosa** [comment]
Thanks @chanseokoh ! Whatever you think is best on where it should be.

**Comment by tejaskasetty** [comment]
@chanseokoh I have made the changes. If its fine I will go ahead with making changes to the test cases.

**Comment by chanseokoh** [comment]
Fixed by #2052. `pullManifest()` will return both `ManifestTemplate` and digest. Note that the digest may be incorrect for Docker Manifest V2 Schema 1, according to https://github.com/GoogleContainerTools/jib/issues/1884#issuecomment-540064001.

**Comment by chanseokoh** [comment]
@aespinosa we've released jib-core 0.12.0, so you can get this now. However, `RegistryClient` isn't a public-facing API.

**Comment by aespinosa** [comment]
thanks! i'll keep note of that.
